File: examples/java/HelloQUAC.java
/**
 * QUAC 100 Hello World Example (Java)
 * 
 * Basic example demonstrating device initialization and random generation.
 * 
 * Build: javac HelloQUAC.java
 * Run:   java HelloQUAC
 * 
 * Copyright 2025 Dyber, Inc. All Rights Reserved.
 */

import java.security.SecureRandom;
import java.util.Arrays;

/**
 * Simulated QUAC Algorithm enum
 */
enum Algorithm {
    ML_KEM_512,
    ML_KEM_768,
    ML_KEM_1024,
    ML_DSA_44,
    ML_DSA_65,
    ML_DSA_87
}

/**
 * Simulated Device Info
 */
class DeviceInfo {
    public final String name;
    public final String serial;
    public final String firmware;
    
    public DeviceInfo(String name, String serial, String firmware) {
        this.name = name;
        this.serial = serial;
        this.firmware = firmware;
    }
}

/**
 * Simulated Key Pair
 */
class KeyPair {
    public final byte[] publicKey;
    public final byte[] secretKey;
    
    public KeyPair(byte[] publicKey, byte[] secretKey) {
        this.publicKey = publicKey;
        this.secretKey = secretKey;
    }
    
    public void clear() {
        Arrays.fill(secretKey, (byte) 0);
    }
}

/**
 * Simulated Device
 */
class Device implements AutoCloseable {
    private final DeviceInfo info;
    private final SecureRandom random = new SecureRandom();
    private boolean closed = false;
    
    public Device(DeviceInfo info) {
        this.info = info;
    }
    
    public DeviceInfo getInfo() {
        return info;
    }
    
    public byte[] random(int length) {
        byte[] bytes = new byte[length];
        random.nextBytes(bytes);
        return bytes;
    }
    
    public KeyPair kemKeygen(Algorithm algorithm) {
        int pkSize, skSize;
        switch (algorithm) {
            case ML_KEM_512:  pkSize = 800;  skSize = 1632; break;
            case ML_KEM_768:  pkSize = 1184; skSize = 2400; break;
            case ML_KEM_1024: pkSize = 1568; skSize = 3168; break;
            default: throw new IllegalArgumentException("Invalid algorithm");
        }
        byte[] pk = new byte[pkSize];
        byte[] sk = new byte[skSize];
        random.nextBytes(pk);
        random.nextBytes(sk);
        return new KeyPair(pk, sk);
    }
    
    @Override
    public void close() {
        closed = true;
    }
    
    public boolean isClosed() {
        return closed;
    }
}

/**
 * Simulated Context
 */
class Context implements AutoCloseable {
    private boolean closed = false;
    
    public int getDeviceCount() {
        return 1;
    }
    
    public Device openDevice(int index) throws Exception {
        throw new Exception("No hardware available");
    }
    
    public Device openSimulator() {
        return new Device(new DeviceInfo(
            "QUAC 100 Simulator",
            "SIM-00000000",
            "1.0.0-sim"
        ));
    }
    
    @Override
    public void close() {
        closed = true;
    }
}

/**
 * Main Hello World Example
 */
public class HelloQUAC {
    
    private static String toHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
    
    public static void main(String[] args) {
        System.out.println("==============================================");
        System.out.println("  QUAC 100 Hello World Example (Java)");
        System.out.println("==============================================");
        System.out.println();
        
        // Step 1: Initialize SDK
        System.out.println("1. Initializing QUAC SDK...");
        try (Context ctx = new Context()) {
            System.out.println("   SDK initialized successfully.");
            System.out.println();
            
            // Step 2: Query devices
            System.out.println("2. Querying devices...");
            int count = ctx.getDeviceCount();
            System.out.println("   Found " + count + " device(s)");
            System.out.println();
            
            // Step 3: Open device
            System.out.println("3. Opening device...");
            Device device;
            try {
                device = ctx.openDevice(0);
            } catch (Exception e) {
                System.out.println("   No hardware found, using simulator...");
                device = ctx.openSimulator();
            }
            
            try {
                System.out.println("   Device opened successfully.");
                System.out.println();
                
                // Step 4: Get device info
                System.out.println("4. Device Information:");
                DeviceInfo info = device.getInfo();
                System.out.println("   Name:     " + info.name);
                System.out.println("   Serial:   " + info.serial);
                System.out.println("   Firmware: " + info.firmware);
                System.out.println();
                
                // Step 5: Generate random bytes
                System.out.println("5. Generating random bytes with QRNG...");
                byte[] randomBytes = device.random(32);
                System.out.println("   Random: " + toHex(randomBytes));
                System.out.println();
                
                // Step 6: Quick ML-KEM-768 demo
                System.out.println("6. Quick ML-KEM-768 demonstration...");
                KeyPair kp = device.kemKeygen(Algorithm.ML_KEM_768);
                System.out.println("   Public key:  " + kp.publicKey.length + " bytes");
                System.out.println("   Secret key:  " + kp.secretKey.length + " bytes");
                System.out.println("   PK (first 16 bytes): " + 
                    toHex(Arrays.copyOf(kp.publicKey, 16)) + "...");
                
                // Secure cleanup
                kp.clear();
                System.out.println();
                
                // Step 7: Cleanup
                System.out.println("7. Cleaning up...");
                
            } finally {
                device.close();
            }
            
            System.out.println("   Done!");
            System.out.println();
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
        
        System.out.println("==============================================");
        System.out.println("  Hello World Complete!");
        System.out.println("==============================================");
    }
}

File: examples/rust/hello_quac.rs
//! QUAC 100 Hello World Example (Rust)
//!
//! Basic example demonstrating device initialization and random generation.
//!
//! Run: cargo run --example hello_quac
//!
//! Copyright 2025 Dyber, Inc. All Rights Reserved.

use rand::RngCore;
use zeroize::Zeroize;

// =============================================================================
// Simulated QUAC Types (would be imported from quac100 crate)
// =============================================================================

#[derive(Debug, Clone, Copy)]
pub enum Algorithm {
    MlKem512,
    MlKem768,
    MlKem1024,
}

#[derive(Debug)]
pub struct DeviceInfo {
    pub name: String,
    pub serial: String,
    pub firmware: String,
}

pub struct Device {
    info: DeviceInfo,
    _is_simulator: bool,
}

impl Device {
    pub fn get_info(&self) -> &DeviceInfo {
        &self.info
    }

    pub fn random(&self, length: usize) -> Vec<u8> {
        let mut buf = vec![0u8; length];
        rand::thread_rng().fill_bytes(&mut buf);
        buf
    }

    pub fn kem_keygen(&self, algorithm: Algorithm) -> (Vec<u8>, Vec<u8>) {
        let (pk_size, sk_size) = match algorithm {
            Algorithm::MlKem512 => (800, 1632),
            Algorithm::MlKem768 => (1184, 2400),
            Algorithm::MlKem1024 => (1568, 3168),
        };

        let mut pk = vec![0u8; pk_size];
        let mut sk = vec![0u8; sk_size];
        rand::thread_rng().fill_bytes(&mut pk);
        rand::thread_rng().fill_bytes(&mut sk);
        (pk, sk)
    }
}

pub struct Context;

impl Context {
    pub fn new() -> Result<Self, &'static str> {
        Ok(Context)
    }

    pub fn get_device_count(&self) -> usize {
        1 // Simulator always available
    }

    pub fn open_device(&self, _index: usize) -> Result<Device, &'static str> {
        Err("No hardware available")
    }

    pub fn open_simulator(&self) -> Result<Device, &'static str> {
        Ok(Device {
            info: DeviceInfo {
                name: "QUAC 100 Simulator".to_string(),
                serial: "SIM-00000000".to_string(),
                firmware: "1.0.0-sim".to_string(),
            },
            _is_simulator: true,
        })
    }
}

// =============================================================================
// Main
// =============================================================================

fn main() {
    println!("==============================================");
    println!("  QUAC 100 Hello World Example (Rust)");
    println!("==============================================");
    println!();

    // Step 1: Initialize SDK
    println!("1. Initializing QUAC SDK...");
    let ctx = Context::new().expect("Failed to initialize SDK");
    println!("   SDK initialized successfully.");
    println!();

    // Step 2: Query devices
    println!("2. Querying devices...");
    let count = ctx.get_device_count();
    println!("   Found {} device(s)", count);
    println!();

    // Step 3: Open device
    println!("3. Opening device...");
    let device = match ctx.open_device(0) {
        Ok(d) => d,
        Err(_) => {
            println!("   No hardware found, using simulator...");
            ctx.open_simulator().expect("Failed to open simulator")
        }
    };
    println!("   Device opened successfully.");
    println!();

    // Step 4: Get device info
    println!("4. Device Information:");
    let info = device.get_info();
    println!("   Name:     {}", info.name);
    println!("   Serial:   {}", info.serial);
    println!("   Firmware: {}", info.firmware);
    println!();

    // Step 5: Generate random bytes
    println!("5. Generating random bytes with QRNG...");
    let random_bytes = device.random(32);
    println!("   Random: {}", hex::encode(&random_bytes));
    println!();

    // Step 6: Quick ML-KEM-768 demo
    println!("6. Quick ML-KEM-768 demonstration...");
    let (pk, mut sk) = device.kem_keygen(Algorithm::MlKem768);
    println!("   Public key:  {} bytes", pk.len());
    println!("   Secret key:  {} bytes", sk.len());
    println!("   PK (first 16 bytes): {}...", hex::encode(&pk[..16]));

    // Secure cleanup
    sk.zeroize();
    println!();

    // Step 7: Cleanup
    println!("7. Cleaning up...");
    println!("   Done!");
    println!();

    println!("==============================================");
    println!("  Hello World Complete!");
    println!("==============================================");
}